File: hicexplorer/chicQualityControl.py
import argparse
import math
from multiprocessing import Process, Queue
import time
import os
import logging
logging.getLogger('hicmatrix').setLevel(logging.CRITICAL)
log = logging.getLogger(__name__)

import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

from hicmatrix import HiCMatrix as hm
from hicexplorer._version import __version__
from .lib import Viewpoint


def parse_arguments(args=None):

    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawDescriptionHelpFormatter,
        add_help=False,
        description="""
Computes the sparsity of each viewpoint to determine the quality. A viewpoint is considered to be of bad quality if it is too sparse i.e. if there are too many locations with no interactions recorded.

This script creates three output files: a plot with the sparsity distribution per matrix, a plot with the sparsity distribution as histograms and a filtered reference points file.

An example usage is:

$ chicQualityControl -m matrix1.cool matrix2.cool -rp referencePointsFile.bed --range 20000 40000 --sparsity 0.01 -o referencePointFile_QC_passed.bed
"""
    )

    parserRequired = parser.add_argument_group('Required arguments')

    parserRequired.add_argument('--matrices', '-m',
                                help='The input matrices to apply the QC on.',
                                nargs='+',
                                required=True)

    parserRequired.add_argument('--referencePoints', '-rp',
                                help='Bed file contains all reference points which are checked for a sufficient number of interactions.',
                                type=str,
                                required=True)
    parserRequired.add_argument('--sparsity', '-s',
                                help='Viewpoints with a sparsity less than the value given are considered of bad quality. If multiple matrices are given, '
                                'the viewpoint is removed as soon as it is of bad quality in at least one matrix.',
                                type=float,
                                required=True)

    parserOpt = parser.add_argument_group('Optional arguments')

    parserOpt.add_argument('--outFileName', '-o',
                           help='The output file name of the passed reference points. Used as prefix for the plots as well'
                           ' (Default: %(default)s).',
                           default='new_referencepoints.bed')
    parserOpt.add_argument('--outFileNameHistogram', '-oh',
                           help='The output file for the histogram plot'
                           ' (Default: %(default)s).',
                           default='histogram.png')
    parserOpt.add_argument('--outFileNameSparsity', '-os',
                           help='The output file for the sparsity distribution plot'
                           ' (Default: %(default)s).',
                           default='sparsity.png')
    parserOpt.add_argument('--threads', '-t',
                           help='Number of threads'
                           ' (Default: %(default)s).',
                           required=False,
                           default=4,
                           type=int
                           )
    parserOpt.add_argument('--fixateRange', '-fs',
                           help='Fixate score of background model starting at distance x. E.g. all values greater than 500kb are set to the value of the 500kb bin'
                           ' (Default: %(default)s).',
                           required=False,
                           default=500000,
                           type=int
                           )
    parserOpt.add_argument('--dpi',
                           help='Optional parameter: Resolution for the image if the'
                           'output is a raster graphics image (e.g png, jpg)'
                           ' (Default: %(default)s).',
                           type=int,
                           default=300,
                           required=False)
    parserOpt.add_argument('--help', '-h', action='help',
                           help='show this help message and exit')

    parserOpt.add_argument('--version', action='version',
                           version='%(prog)s {}'.format(__version__))

    return parser


def compute_sparsity(pReferencePoints, pViewpointObj, pArgs, pQueue):

    sparsity_list = []
    try:
        chromosome_names = pViewpointObj.hicMatrix.getChrNames()

        for i, referencePoint in enumerate(pReferencePoints):

            if referencePoint is not None and referencePoint[0] in chromosome_names:

                region_start, region_end, _ = pViewpointObj.calculateViewpointRange(
                    referencePoint, (pArgs.fixateRange, pArgs.fixateRange))
                try:
                    data_list, _ = pViewpointObj.computeViewpoint(
                        referencePoint, referencePoint[0], region_start, region_end)
                    sparsity = (np.count_nonzero(data_list) / len(data_list))
                except (TypeError, IndexError):
                    sparsity = -1.0
                sparsity_list.append(sparsity)
            else:
                sparsity_list.append(-1.0)
    except Exception as exp:
        pQueue.put('Fail: ' + str(exp))
        return

    pQueue.put(sparsity_list)
    return


def main(args=None):
    args = parse_arguments().parse_args(args)
    matplotlib.rcParams['pdf.fonttype'] = 42
    viewpointObj = Viewpoint()
    referencePoints, _ = viewpointObj.readReferencePointFile(
        args.referencePoints)

    # compute for each viewpoint the sparsity and consider these as bad with a sparsity less than given.

    referencePointsPerThread = len(referencePoints) // args.threads
    queue = [None] * args.threads
    process = [None] * args.threads
    sparsity = []
    fail_flag = False
    fail_message = ''
    for j, matrix in enumerate(args.matrices):
        sparsity_local = [None] * args.threads
        hic_ma = hm.hiCMatrix(matrix)
        viewpointObj.hicMatrix = hic_ma

        all_data_collected = False
        thread_done = [False] * args.threads
        for i in range(args.threads):

            if i < args.threads - 1:
                referencePointsThread = referencePoints[i * referencePointsPerThread:(i + 1) * referencePointsPerThread]
            else:
                referencePointsThread = referencePoints[i * referencePointsPerThread:]
            if len(referencePointsThread) == 0:
                process[i] = None
                queue[i] = None
                sparsity_local[i] = []
                continue
            else:
                queue[i] = Queue()
                process[i] = Process(target=compute_sparsity, kwargs=dict(
                    pReferencePoints=referencePointsThread,
                    pViewpointObj=viewpointObj,
                    pArgs=args,
                    pQueue=queue[i]
                )
                )

                process[i].start()
                log.debug('process started {}'.format(i))

        while not all_data_collected:
            for i in range(args.threads):
                if queue[i] is not None and not queue[i].empty():
                    sparsity_ = queue[i].get()
                    if 'Fail:' in sparsity_:
                        fail_flag = True
                        fail_message = sparsity_[6:]
                    log.debug('process computed: {}'.format(i))
                    sparsity_local[i] = sparsity_
                    queue[i] = None
                    process[i].join()
                    process[i].terminate()
                    process[i] = None
                    thread_done[i] = True
            all_data_collected = True
            for thread in thread_done:
                if not thread:
                    all_data_collected = False
            time.sleep(1)

        del hic_ma
        del viewpointObj.hicMatrix

        # merge sparsity data per matrix from each thread to one list
        if fail_flag:
            log.error(fail_message)
            exit(1)
        sparsity_local = [item for sublist in sparsity_local for item in sublist]
        sparsity.append(sparsity_local)

    # change sparsity to sparsity values per viewpoint per matrix: viewpoint = [matrix1, ..., matrix_n]
    sparsity = np.array(sparsity).T
    count_accepted = 0
    count_rejected = 0
    count_failure = 0
    with open(args.referencePoints, 'r') as reference_file_input:
        with open(args.outFileName + '_raw_filter', 'w') as output_file_raw:
            output_file_raw.write('# Created with chicQualityControl version {}\n'.format(__version__))
            output_file_raw.write('# A sparsity of -1.0 indicates a faulty reference point e.g. no data for this reference point was in the matrix.\n')
            output_file_raw.write('# Used Matrices ')
            for matrix in args.matrices:
                output_file_raw.write('{}\t'.format(matrix))
            output_file_raw.write('\n# Chromosome\tStart\tEnd')
            for matrix in args.matrices:
                output_file_raw.write('\tSparsity {}'.format(os.path.basename(matrix)))
            output_file_raw.write('\n')

            with open(args.outFileName + '_failed_reference_points', 'w') as output_file_failed:
                with open(args.outFileName + '_rejected_filter', 'w') as output_file_rejected:
                    with open(args.outFileName, 'w') as output_file:
                        for i, line in enumerate(reference_file_input.readlines()):
                            sparsity_str = '\t'.join(str(x) for x in sparsity[i])
                            output_file_raw.write(
                                line.strip() + '\t' + sparsity_str + '\n')
                            count = 0
                            count_negative = 0
                            for j in range(len(sparsity[i])):
                                if sparsity[i][j] == -1.0:
                                    count_negative += 1
                                elif sparsity[i][j] > args.sparsity:
                                    count += 1
                            if count_negative:
                                output_file_failed.write(line)
                                count_failure += 1
                            elif count:
                                output_file.write(line)
                                count_accepted += 1
                            else:
                                output_file_rejected.write(line)
                                count_rejected += 1

    with open(args.outFileName + '_report', 'w') as output_file_report:
        output_file_report.write('# Created with chicQualityControl version {}\n'.format(__version__))
        output_file_report.write('# QC report for matrices: ')
        for matrix in args.matrices:
            output_file_report.write(matrix + ' ')
        output_file_report.write('\n')
        output_file_report.write('#Sparsity threshold for rejection: sparsity <= {} are rejected.\n'.format(args.sparsity))
        output_file_report.write('\nNumber of reference points: {}\n'.format(str(count_accepted + count_rejected + count_failure)))
        output_file_report.write('Number of accepted reference points: {}\n'.format(str(count_accepted)))
        output_file_report.write('Number of rejected reference points: {}\n'.format(str(count_rejected)))
        output_file_report.write('Number of faulty reference points: {}\n'.format(str(count_failure)))
        output_file_report.write('\n\nA faulty reference point is caused by the non-presence of the chromosome in one of the given matrices.\n')
        output_file_report.write('It can also be caused by the non-presence of valid Hi-C reads in a region, especially at the chromosome ends.\n')
        output_file_report.write('Please check the results of hicInfo to validate this for your data.\n')

    # output plot of sparsity distribution per sample
    # remove fault reference points from statistics
    x = [[]] * len(args.matrices)
    y = [[]] * len(args.matrices)

    mask = [True] * len(sparsity)
    for i in range(len(sparsity)):
        delete_instance = False
        for j in range(len(args.matrices)):
            if sparsity[i][j] == -1.0:
                delete_instance = True
        if delete_instance:
            mask[i] = False

    mask = np.array(mask)
    sparsity = sparsity[mask]

    for i in range(len(args.matrices)):
        y[i] = [i] * len(sparsity)

    sparsity = sparsity.T

    for i in range(len(args.matrices)):
        x[i] = sparsity[i].flatten()

    for i in range(len(args.matrices)):
        plt.plot(x[i], y[i], 'o', mfc='none', markersize=0.3,
                 label=args.matrices[i].split('/')[-1])
    plt.yticks([])
    plt.xlabel("Sparsity level")

    plt.axvline(x=args.sparsity, c='r', label='sparsity threshold', linewidth=0.3)
    plt.xscale('log')
    ax = plt.gca()
    box = ax.get_position()
    ax.set_position([box.x0, box.y0, box.width * 0.6, box.height])
    plt.legend(loc='center', bbox_to_anchor=(1.4, 0.5))
    plt.savefig(args.outFileNameSparsity, dpi=args.dpi)

    plt.close()
    for i in range(len(args.matrices)):
        plt.hist(x[i], bins=100, alpha=0.5, label=args.matrices[i].split('/')[-1])
    plt.xlabel("Sparsity level")
    plt.ylabel("Number of counts")

    ax = plt.gca()
    box = ax.get_position()
    ax.set_position([box.x0, box.y0, box.width * 0.6, box.height])
    plt.legend(loc='center', bbox_to_anchor=(1.4, 0.5))
    plt.savefig(args.outFileNameHistogram, dpi=args.dpi)

File: cpp/tools/chicQualityControl.cpp
// Port of hicexplorer/chicQualityControl.py.
//
// For every reference point and every matrix: the fraction of non zero
// elements in the viewpoint over --fixateRange on either side. A reference
// point is written to --outFileName when it passes the threshold in at least
// one matrix, to *_rejected_filter when it passes in none, and to
// *_failed_reference_points when it is faulty (sparsity -1.0) in any.
//
// Behaviour reproduced as the Python has it, not as its help text says:
//
//   * --sparsity is documented as removing a viewpoint as soon as it is of bad
//     quality in at least one matrix, but chicQualityControl.py:228-240 accepts
//     it when any matrix is above the threshold. Reported as a defect.
//   * The output loop re-reads the reference point file line by line and
//     indexes the sparsity table with the line number. A line the reader
//     skipped (a blank line, or one with fewer than three fields) therefore
//     shifts every later line onto the wrong sparsity, and once the line count
//     exceeds the table the Python raises IndexError after having written the
//     earlier lines. Reproduced: the same lines are written, then exit 1.
//   * A reference point on a chromosome the matrix has, but outside its bins,
//     is a TypeError inside computeViewpoint and becomes -1.0. A reference
//     point whose start lies after its end is a ValueError there, which is not
//     caught, and aborts the run with exit 1.
//
// Figures (cpp/PLAN.md tier 7, option (a)). The Python also draws
// sparsity.png and histogram.png with matplotlib (chicQualityControl.py:262-307).
// After the report the sparsity of every reference point that is faulty in
// no matrix goes, per matrix, to plot/hicexplorer_plot/chicQualityControl.py,
// which draws both figures with the reference's calls. The C++-only option
// --plotData writes that data as JSON instead.
//
// Threading: measured, and not used. Spreading the 45 reference points of
// the test data over worker threads does not shorten the run (0.090 s of wall
// time at --threads 1, 0.093 s at 16) but raises the CPU time from 0.080 s to
// 0.377 s and the peak RSS from 45 MB to 120 MB, above this tool's memory
// budget: the work per reference point is a few thousand sparse lookups, and
// every worker thread brings its own malloc arena. Per cpp/OPTIMIZATION.md 6
// the loop is sequential. --threads is accepted, and 0 still fails as the
// Python's integer division by it does. The output is written in file order;
// the Python collects its worker chunks in index order, which is the same
// order, so neither output depends on the thread count.

#include <cstdio>
#include <exception>
#include <fstream>
#include <string>
#include <vector>

#include "hicx/argparse.hpp"
#include "hicx/chic_viewpoint.hpp"
#include "hicx/numpy_compat.hpp"
#include "hicx/plot_bridge.hpp"
#include "hicx/version.hpp"

namespace {

using hicx::chic::ReferencePoint;

const char* const kUsage =
    "usage: chicQualityControl --matrices MATRICES [MATRICES ...] --referencePoints\n"
    "                          REFERENCEPOINTS --sparsity SPARSITY\n"
    "                          [--outFileName OUTFILENAME]\n"
    "                          [--outFileNameHistogram OUTFILENAMEHISTOGRAM]\n"
    "                          [--outFileNameSparsity OUTFILENAMESPARSITY]\n"
    "                          [--threads THREADS] [--fixateRange FIXATERANGE]\n"
    "                          [--dpi DPI] [--help] [--version]\n";

const char* const kHelp =
    "\n"
    "Computes the sparsity of each viewpoint to determine the quality. A viewpoint is "
    "considered to be of bad quality if it is too sparse i.e. if there are too many "
    "locations with no interactions recorded.\n"
    "\n"
    "This script creates three output files: a plot with the sparsity distribution per "
    "matrix, a plot with the sparsity distribution as histograms and a filtered reference "
    "points file.\n"
    "\n"
    "An example usage is:\n"
    "\n"
    "$ chicQualityControl -m matrix1.cool matrix2.cool -rp referencePointsFile.bed --range "
    "20000 40000 --sparsity 0.01 -o referencePointFile_QC_passed.bed\n"
    "\n"
    "Required arguments:\n"
    "  --matrices MATRICES [MATRICES ...], -m MATRICES [MATRICES ...]\n"
    "                        The input matrices to apply the QC on.\n"
    "  --referencePoints REFERENCEPOINTS, -rp REFERENCEPOINTS\n"
    "                        Bed file contains all reference points which are\n"
    "                        checked for a sufficient number of interactions.\n"
    "  --sparsity SPARSITY, -s SPARSITY\n"
    "                        Viewpoints with a sparsity less than the value given\n"
    "                        are considered of bad quality. If multiple matrices\n"
    "                        are given, the viewpoint is removed as soon as it is\n"
    "                        of bad quality in at least one matrix.\n"
    "\n"
    "Optional arguments:\n"
    "  --outFileName OUTFILENAME, -o OUTFILENAME\n"
    "                        The output file name of the passed reference points.\n"
    "                        Used as prefix for the plots as well (Default:\n"
    "                        new_referencepoints.bed).\n"
    "  --outFileNameHistogram OUTFILENAMEHISTOGRAM, -oh OUTFILENAMEHISTOGRAM\n"
    "                        The output file for the histogram plot (Default:\n"
    "                        histogram.png).\n"
    "  --outFileNameSparsity OUTFILENAMESPARSITY, -os OUTFILENAMESPARSITY\n"
    "                        The output file for the sparsity distribution plot\n"
    "                        (Default: sparsity.png).\n"
    "  --threads THREADS, -t THREADS\n"
    "                        Number of threads (Default: 4).\n"
    "  --fixateRange FIXATERANGE, -fs FIXATERANGE\n"
    "                        Fixate score of background model starting at distance\n"
    "                        x. E.g. all values greater than 500kb are set to the\n"
    "                        value of the 500kb bin (Default: 500000).\n"
    "  --dpi DPI             Optional parameter: Resolution for the image if\n"
    "                        theoutput is a raster graphics image (e.g png, jpg)\n"
    "                        (Default: 300).\n"
    "  --help, -h            show this help message and exit\n"
    "  --version             show program's version number and exit\n"
    "\n"
    "C++ port: the sparsity and the reference point files are computed in C++, and\n"
    "the two figures are drawn by the hicexplorer_plot drawing layer with the\n"
    "matplotlib calls of the Python tool (HICX_PLOT_PYTHON names the interpreter).\n"
    "The C++-only option --plotData FILE writes the data of the figures as JSON to\n"
    "FILE instead of drawing them.\n";

std::string basename(const std::string& path) {
    const std::size_t slash = path.find_last_of('/');
    return slash == std::string::npos ? path : path.substr(slash + 1);
}

struct Slot {
    double sparsity = -1.0;
    bool failed = false;
    std::string message;
};

}  // namespace

int main(int argc, char** argv) {
    namespace cli = hicx::cli;
    cli::Parser parser("chicQualityControl",
                       "Computes the sparsity of each viewpoint to determine the quality.");
    parser.set_usage(kUsage).set_help(kHelp).set_version_string(hicx::kVersion);
    cli::ArgumentGroup& required = parser.group("Required arguments");
    required.add({"--matrices", "-m"})
        .nargs("+")
        .required()
        .input({"cool", "h5"})
        .help("The input matrices to apply the QC on.");
    required.add({"--referencePoints", "-rp"})
        .type("str")
        .required()
        .input({"bed"})
        .help("Bed file contains all reference points which are checked for a sufficient number "
              "of interactions.");
    required.add({"--sparsity", "-s"})
        .type("float")
        .required()
        .help("Viewpoints with a sparsity less than the value given are considered of bad "
              "quality.");
    cli::ArgumentGroup& optional = parser.group("Optional arguments");
    optional.add({"--outFileName", "-o"})
        .default_value("new_referencepoints.bed")
        .output({"bed"})
        .help("The output file name of the passed reference points.");
    optional.add({"--outFileNameHistogram", "-oh"})
        .default_value("histogram.png")
        .output({"png", "pdf", "svg"})
        .help("The output file for the histogram plot");
    optional.add({"--outFileNameSparsity", "-os"})
        .default_value("sparsity.png")
        .output({"png", "pdf", "svg"})
        .help("The output file for the sparsity distribution plot");
    optional.add({"--threads", "-t"}).type("int").default_value(4).help("Number of threads");
    optional.add({"--fixateRange", "-fs"})
        .type("int")
        .default_value(500000)
        .help("Range on either side of a reference point the sparsity is computed over.");
    optional.add({"--dpi"})
        .type("int")
        .default_value(300)
        .help("Resolution for the image if the output is a raster graphics image.");
    optional.add({"--help", "-h"}).action(cli::Action::Help).help("show this help message and exit");
    optional.add({"--version"}).version(std::string("%(prog)s ") + hicx::kVersion);
    optional.add({"--plotData"})
        .metavar("FILE")
        .output({"json"})
        .cpp_only("The data of the figures as JSON, without drawing them (cpp/PLAN.md tier 7).")
        .help("Write the data the two figures are drawn from as JSON to this file and do not "
              "draw them.");
    const cli::Namespace args = parser.parse(argc, argv);
    if (const int refused = hicx::plot::preflight("chicQualityControl", !args.given("plotData")); refused != 0) {
        return refused;
    }
    std::string plot_json;

    const std::vector<std::string>& matrices = args.strs("matrices");
    const std::string reference_point_file = args.str("referencePoints");
    const double sparsity_threshold = args.real("sparsity");
    const std::string out_file_name = args.str("outFileName");
    const std::int64_t threads = args.integer("threads");
    const std::int64_t fixate_range = args.integer("fixateRange");

    try {
        const hicx::chic::ReferencePoints reference_points =
            hicx::chic::read_reference_points(reference_point_file);
        const std::size_t count = reference_points.points.size();
        if (threads == 0) {
            // len(referencePoints) // args.threads
            throw std::runtime_error("ZeroDivisionError: integer division or modulo by zero");
        }

        // sparsity[matrix][reference point]
        std::vector<std::vector<double>> sparsity;
        for (const std::string& matrix_path : matrices) {
            const hicx::chic::ViewpointMatrix matrix = hicx::chic::ViewpointMatrix::load(matrix_path);
            const std::vector<std::string> names = matrix.chromosome_names();
            std::vector<Slot> slots(count);
            const auto compute_slot = [&](std::size_t index) {
                Slot& slot = slots[index];
                const ReferencePoint& point = reference_points.points[index];
                try {
                    if (!matrix.has_chromosome(point.chromosome)) {
                        slot.sparsity = -1.0;
                        return;
                    }
                    const hicx::chic::ViewpointRange range = hicx::chic::calculate_viewpoint_range(
                        matrix, point, fixate_range, fixate_range);
                    try {
                        const hicx::chic::ComputedViewpoint viewpoint = hicx::chic::compute_viewpoint(
                            matrix, point, point.chromosome, range.region_start, range.region_end);
                        std::size_t nonzero = 0;
                        for (const double value : viewpoint.data) {
                            nonzero += value != 0.0 ? 1 : 0;
                        }
                        if (viewpoint.data.empty()) {
                            throw std::runtime_error("ZeroDivisionError: division by zero");
                        }
                        slot.sparsity = static_cast<double>(nonzero) /
                                        static_cast<double>(viewpoint.data.size());
                    } catch (const hicx::chic::TypeError&) {
                        slot.sparsity = -1.0;
                    } catch (const hicx::chic::IndexError&) {
                        slot.sparsity = -1.0;
                    }
                } catch (const std::exception& error) {
                    slot.failed = true;
                    slot.message = error.what();
                }
            };
            for (std::size_t index = 0; index < count; ++index) {
                compute_slot(index);
            }
            for (const Slot& slot : slots) {
                if (slot.failed) {
                    std::fprintf(stderr, "chicQualityControl: %s\n", slot.message.c_str());
                    return 1;
                }
            }
            std::vector<double> column(count);
            for (std::size_t i = 0; i < count; ++i) {
                column[i] = slots[i].sparsity;
            }
            sparsity.push_back(std::move(column));
        }

        const std::vector<std::string> lines = hicx::chic::read_lines(reference_point_file);
        std::size_t accepted = 0;
        std::size_t rejected = 0;
        std::size_t failures = 0;
        {
            std::ofstream raw(out_file_name + "_raw_filter", std::ios::binary);
            raw << "# Created with chicQualityControl version " << hicx::kVersion << "\n";
            raw << "# A sparsity of -1.0 indicates a faulty reference point e.g. no data for "
                   "this reference point was in the matrix.\n";
            raw << "# Used Matrices ";
            for (const std::string& matrix_path : matrices) {
                raw << matrix_path << "\t";
            }
            raw << "\n# Chromosome\tStart\tEnd";
            for (const std::string& matrix_path : matrices) {
                raw << "\tSparsity " << basename(matrix_path);
            }
            raw << "\n";

            std::ofstream failed(out_file_name + "_failed_reference_points", std::ios::binary);
            std::ofstream rejected_file(out_file_name + "_rejected_filter", std::ios::binary);
            std::ofstream accepted_file(out_file_name, std::ios::binary);
            for (std::size_t i = 0; i < lines.size(); ++i) {
                if (i >= count) {
                    std::fprintf(stderr,
                                 "chicQualityControl: IndexError: index %zu is out of bounds for "
                                 "axis 0 with size %zu\n",
                                 i, count);
                    return 1;
                }
                std::string sparsity_text;
                std::size_t passing = 0;
                std::size_t negative = 0;
                for (std::size_t j = 0; j < sparsity.size(); ++j) {
                    const double value = sparsity[j][i];
                    sparsity_text += (j == 0 ? "" : "\t") + hicx::npy::float_repr(value);
                    if (value == -1.0) {
                        ++negative;
                    } else if (value > sparsity_threshold) {
                        ++passing;
                    }
                }
                raw << hicx::chic::strip(lines[i]) << "\t" << sparsity_text << "\n";
                if (negative > 0) {
                    failed << lines[i];
                    ++failures;
                } else if (passing > 0) {
                    accepted_file << lines[i];
                    ++accepted;
                } else {
                    rejected_file << lines[i];
                    ++rejected;
                }
            }
        }

        {
            std::ofstream report(out_file_name + "_report", std::ios::binary);
            report << "# Created with chicQualityControl version " << hicx::kVersion << "\n";
            report << "# QC report for matrices: ";
            for (const std::string& matrix_path : matrices) {
                report << matrix_path << " ";
            }
            report << "\n";
            report << "#Sparsity threshold for rejection: sparsity <= "
                   << hicx::npy::float_repr(sparsity_threshold) << " are rejected.\n";
            report << "\nNumber of reference points: " << (accepted + rejected + failures) << "\n";
            report << "Number of accepted reference points: " << accepted << "\n";
            report << "Number of rejected reference points: " << rejected << "\n";
            report << "Number of faulty reference points: " << failures << "\n";
            report << "\n\nA faulty reference point is caused by the non-presence of the "
                      "chromosome in one of the given matrices.\n";
            report << "It can also be caused by the non-presence of valid Hi-C reads in a "
                      "region, especially at the chromosome ends.\n";
            report << "Please check the results of hicInfo to validate this for your data.\n";
        }

        // chicQualityControl.py:262-281: the reference points faulty in no
        // matrix, then x[i] = the sparsity column of matrix i.
        std::vector<std::string> labels;
        std::vector<std::string> columns;
        for (std::size_t j = 0; j < sparsity.size(); ++j) {
            std::vector<double> kept;
            for (std::size_t i = 0; i < count; ++i) {
                bool faulty = false;
                for (const auto& other : sparsity) {
                    faulty = faulty || other[i] == -1.0;
                }
                if (!faulty) {
                    kept.push_back(sparsity[j][i]);
                }
            }
            columns.push_back(hicx::plot::json_numbers(kept));
            labels.push_back(basename(matrices[j]));
        }
        hicx::plot::JsonObject data;
        data.add("outFileNameSparsity", hicx::plot::json_string(args.str("outFileNameSparsity")));
        data.add("outFileNameHistogram",
                 hicx::plot::json_string(args.str("outFileNameHistogram")));
        data.add("dpi", hicx::plot::json_int(args.integer("dpi")));
        data.add("sparsity", hicx::plot::json_number(sparsity_threshold));
        data.add("labels", hicx::plot::json_strings(labels));
        data.add("x", hicx::plot::json_list(columns));
        plot_json = data.str();
    } catch (const std::exception& error) {
        std::fprintf(stderr, "chicQualityControl: %s\n", error.what());
        return 1;
    }

    return hicx::plot::draw("chicQualityControl", plot_json, args.opt_str("plotData"));
}
